activemodel/lib/active_model/errors.rb

```diff
@@ -258,13 +303,30 @@ def as_json(options = nil)
     #   person.errors.to_hash       # => {:name=>["cannot be nil"]}
     #   person.errors.to_hash(true) # => {:name=>["name cannot be nil"]}
     def to_hash(full_messages = false)
-      if full_messages
-        messages.each_with_object({}) do |(attribute, array), messages|
-          messages[attribute] = array.map { |message| full_message(attribute, message) }
-        end
-      else
-        without_default_proc(messages)
+      hash = {}
+      message_method = full_messages ? :full_message : :message
+      group_by_attribute.each do |attribute, errors|
+        hash[attribute] = errors.map(&message_method)
+      end
+      hash
+    end
+
+    def messages
+      DeprecationHandlingMessageHash.new(self)
+    end
+
+    def details
+      hash = {}
+      group_by_attribute.each do |attribute, errors|
+        hash[attribute] = errors.map { |error|
+          normalize_detail(error.type, error.options)
```

Comment:
When you add an error with a message `errors.add(:base, "Foo")` before defailts returned `error: "Foo"` now it is returning `error: :invalid`. We should keep that behavior.

Reply:
The new design are build around the fundamental that `type` has to be present, and be a `:symbol`. I feel this is a breaking change that is really needed, so could we keep this (since this is a major release).

Reply:
I get that type needs to be always present but this is changing the return of `details` without warning. I don't think one is necessary for the other and without deprecating I don't think we can just change that or this will silently break people application. Would it possible to add a new method called `type_or_message` or something similar that does return the thing `details` used to return here?

Reply:
Thanks for the suggestion and I've made the change.

However how would one go about deprecate one field in the detailed hash?

Reply:
If we need to, what I'm not sure yet we can return a subclass of hash that shows deprecation when one key is accessed.

Reply:
I will do the deprecation in a separate PR as it is less urgent and I am occupied for the next two weeks. Thanks!